SWE L1B implementation and updates to L1A

tech3371 opened the pull request:
# Change Summary
Implements SWE L1B algorithm and updates L1A
## Overview
These are the changes in this PR    
    - added logics to populate data based on ESA lookup table number
    - refactored code to store l1a data as 180x7 instead of 15x12x7 array.
    - added new event test data and moved tests data into folders.
    - updated l1a code to be ISTP compliant and to be able to create cdf file
    - added new test data and update test as needed
    - implemented SWE l1b algorithm for both science and other data types
    - creates cdf file for SWE l1b data product
    - added deadtime correction and place holder for in-flight calibration. converted counts to rates
    - added rest of l1b algorithm to convert raw data to engineering unit.

## New Files
- imap_processing/swe/l1b/
   - Added code files to process data to l1b
- imap_processing/swe/tests/
   - Added new tests for l1b and new l1a test data

## Deleted Files
<!--List each deleted file and add one or more bullets with the rationale for why the file is being deleted-->
- deleted file 1
   - explanation for why file was deleted

## Updated Files
- imap_processing/swe/l1a/
   - updated l1a code to be able to create CDF file


## Testing
- imap_processing/swe/tests/
   - Added new tests for l1b and new l1a and l1b test data

tech3371 committed:
SWE L1B implementation and updates to L1A:

- added logics to populate data based on ESA lookup table number
- refactored code to store l1a data as 180x7 instead of 15x12x7 array.
- added new event test data and moved tests data into folders.
- updated l1a code to be ISTP compliant and to be able to create cdf file
- added new test data and update test as needed
- implemented SWE l1b algorithm for both science and other data types
- creates cdf file for SWE l1b data product
- added deadtime correction and place holder for in-flight calibration. converted counts to rates
- added rest of l1b algorithm to convert raw data to engineering unit.

greglucas commented on imap_processing/swe/l1a/swe_l1a.py:
```diff
@@ -1,10 +1,90 @@
+    metadata_arrays = collections.defaultdict(list)
+
+    for data_packet in packets:
+        add_metadata_to_array(data_packet, metadata_arrays)
```
mutating the arguments is potentially ambiguous here. I'd return the metadata_arrays back out of the function to avoid relying on the implicit mutation of the inputs.

```python
metadata_arrays = add_metadata_to_array(data_packet, metadata_arrays)
```

tech3371 replied:
I tried to do that. It seems to gives me back this error. Do you know way around it?
```
E           AttributeError: 'NoneType' object has no attribute 'setdefault'
```

greglucas replied:
I think you need to return the `metadata_arrays` out of the function.

greglucas commented on imap_processing/swe/l1a/swe_l1a.py:
```diff
@@ -13,15 +93,23 @@ def swe_l1a(packet_file: str):
-    else:
-        return decom_data
+    # group data by appId
+    grouped_data = group_by_apid(packets)
```
I think you should be able to test this with the current data you've got? Just assert that the group is in sorted order or something else quick.

tech3371 replied:
I think there is test for this function. Did you like me to add specific test in there?

greglucas replied:
I think this comment was before you added the test and codecov was failing. Resolving this now, nothing needed here.

greglucas commented on imap_processing/swe/l1a/swe_l1a.py:
```diff
@@ -13,15 +93,23 @@ def swe_l1a(packet_file: str):
+            sorted_packets = sorted(
+                grouped_data[apid], key=lambda x: x.data["ACQ_START_COARSE"].raw_value
+            )
+            logging.info("Processing science data")
```
```suggestion
            logging.debug("Processing science data for [%s] packets", len(sorted_packets))
```
Add some more information to the log to help debug later? Also make it a debug log since we don't need it every time and likely only when we are trying to investigate something.

greglucas commented on imap_processing/swe/l1a/swe_l1a.py:
```diff
@@ -13,15 +93,23 @@ def swe_l1a(packet_file: str):
+            # TODO: write data to CDF.
+        else:
+            # If it's not science, we unpack, organize and save it as a dataset.
+            create_dataset(packets=grouped_data[apid])
```
Do you need to access the returned dataset? ds = create_dataset(...)

greglucas commented on imap_processing/swe/l1a/swe_science.py:
```diff
@@ -133,18 +133,18 @@ def swe_science(decom_data):
         # decompression_table and then reshape the return
-        uncompress_data = np.take(decompression_table, raw_counts).reshape(15, 12, 7)
-        raw_counts = raw_counts.reshape(15, 12, 7)
+        uncompress_data = np.take(decompression_table, raw_counts).reshape(180, 7)
+        raw_counts = raw_counts.reshape(180, 7)
```
Update the comment above to indicate this is 180 now? Leave the 15 x 12 information though since that still seems useful.

greglucas commented on imap_processing/swe/l1b/swe_l1b.py:
```diff
@@ -0,0 +1,74 @@
+        ].values[0][6:][::-1]
+
+        # Convert the raw value to engineering unit
+        dataset[metadata_field].data = np.polyval(
+            coeff_values, dataset[metadata_field].data
+        )
```
Use the `Polynomial` class instead because polyval is discouraged. It looks like it is also in increasing order for the coefficients.

https://numpy.org/doc/stable/reference/generated/numpy.polynomial.polynomial.Polynomial.html#numpy.polynomial.polynomial.Polynomial

greglucas commented on imap_processing/swe/l1b/swe_l1b.py:
```diff
@@ -0,0 +1,74 @@
+    # for each metadata field, convert raw value to engineering unit
+    for field in metadata_list.index:
+        metadata_field = metadata_list.loc[field]["mnemonic"]
+        # On this line, we are getting the coefficients from the
```
Great comments here. I'd even add more like:

Data from the tables are the coefficients of polynomials looking like: p0 + p1*x + p2*x**2 which are provided by the instrument teams in calibration files.

greglucas commented on imap_processing/swe/tests/test_swe_l1a.py:
```diff
@@ -0,0 +1,47 @@
+
+    # check total dataset for swe science
+    total_science_data = grouped_data[SWEAPID.SWE_SCIENCE.value]
+    # print(total_science_data)
```
```suggestion
```

greglucas commented on imap_processing/swe/tests/test_swe_l1a_event_msg.py:
```diff
@@ -0,0 +1,49 @@
+    event_file_list = list(
+        imap_module_directory.glob(f"{test_folder_path}/*SWE_EVTMSG_packet.bin")
+    )
+    event_data = []
+    for packet_file in event_file_list:
+        event_data.append(decom_swe.decom_packets(packet_file)[0])
+
+    return event_data
```
Is this comment accurate? I wasn't entirely clear why you were doing this in just code at first.
```suggestion
    # Return a list of the first packet from every file
    return [decom_swe.decom_packets(packet_file)[0] for packet_file in event_file_list]
```

tech3371 replied:
no. I got event data in multiple files. And each file contains one event data only. In this for loop, I am going through all event data file and reading those one data and return as one for testing.

greglucas commented on imap_processing/swe/tests/test_swe_l1a_science.py:
```diff
@@ -18,10 +18,10 @@ def decom_test_data():
     # Packet 4 has spin 3's data
     # reorder to match the order of spin in the data
```
Do we need to be able to account for this re-ordering in code and make sure we can handle this ourselves? Or this is a one-off thing due to the way they ran their simulator?

tech3371 replied:
This is a one-off thing

greglucas commented on imap_processing/swe/tests/test_swe_l1b.py:
```diff
@@ -0,0 +1,92 @@
+        "THRESHOLD_DAC",
+    ]
+
+    # Test EU values for science data
+    for field in science_eu_field_list:
+        assert round(science_l1b[field].data[1], 5) == round(validation_data[field], 5)
```
Would we be able to use numpys testing routines for this at all instead of rounding things yourself, and possibly removing the loop too then.
https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html#numpy.testing.assert_allclose

tech3371 replied:
btw, I looked into this few days ago. I couldn't get it to work without complicating logics to remove the loop. The reason was that, I only check float value matches for only these three variables. If we remember correctly, I had to look up these values from both `science_l1b` and `validation_data`, then match them using `np`. It was looking unreadable and since it's just these three variables, I thought for loop might be ok.
```
science_eu_field_list = [
        "SPIN_PHASE",
        "SPIN_PERIOD",
        "THRESHOLD_DAC",
    ]
```

greglucas commented on imap_processing/swe/tests/test_swe_l1b_science.py:
```diff
@@ -0,0 +1,56 @@
+def test_get_full_cycle_data_indices():
+    q = np.array([0, 1, 2, 0, 1, 2, 3, 2, 3, 0, 2, 3, 0, 1, 2, 3, 2, 3, 1, 0])
+    filtered_q = get_indices_of_full_cycles(q)
+    assert np.array_equal(filtered_q, np.array([3, 4, 5, 6, 12, 13, 14, 15]))
```
Here you could use `np.testing.assert_array_equal()` instead of needing to assert yourself here.

tech3371 committed:
Added project wide utils function and swe specific utils

tech3371 committed:
adding swe_cdf_attrs.py

tech3371 committed:
added argument to ignore too many function argument error in pre-commit hooks. Will remove it once I refactor in upcoming PRs.

tech3371 committed:
import fixes and fix format

tech3371 commented:
pre-commit.ci autofix

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

tech3371 committed:
fixed spelling of conversion

tech3371 committed:
fixing unittest. files may have been read in different in different order. Added checks to look for right data

tech3371 committed:
feedback changes

GFMoraga commented on imap_processing/swe/l1a/swe_science.py:
```diff
@@ -130,50 +118,80 @@ def swe_science(decom_data):
-            units="seconds since start of the mission",
-        ),
+        name="Epoch",
+        dims=["Epoch"],
```
Should this be "EPOCH" in caps?

tech3371 replied:
I had the same question at one point :) After looking at Bryan's code and ISTP, it seems like ISTP likes it `Epoch`.

tech3371 committed:
moved ignore into per-file-ignores

greglucas commented on imap_processing/swe/l1a/swe_l1a.py:
```diff
@@ -13,15 +21,51 @@ def swe_l1a(packet_file: str):
+
+    for apid in grouped_data.keys():
+        # If appId is science, then the file should contain all data of science appId
+        if apid == SWEAPID.SWE_SCIENCE.value:
```
```suggestion
        if apid == SWEAPID.SWE_SCIENCE:
```
since you're using an IntEnum I don't think you need the value here and below.

tech3371 replied:
interesting. I didn't know that.

greglucas commented on imap_processing/swe/l1a/swe_l1a.py:
```diff
@@ -1,10 +1,18 @@
+    get_descriptor,
+)
+from imap_processing.utils import group_by_apid, sort_by_time
+from imap_processing.write_to_cdf import write_to_cdf
```
Can you change this to import from `imap_processing.cdf_utils` now that it is in? Then you could remove this portion and avoid the duplication now instead of having to fix it later.

tech3371 replied:
I started working on it but it introduced more issues. I will revisit this on Monday.

tech3371 replied:
I think I worked out most of issue that day and forgot to update here(because I was sick) but I will double check when I rebase with new changes from Maxine's.

GFMoraga reviewed (approved):
I would maybe double check the `dims` and make sure your time variables don't have to be CAPS as well.  You are most likely doing the correct format. I was just wondering about if the time needs to be CAPS if SHCOARSE is?
 This is an example:
`epoch_time = xr.DataArray(
        metadata_arrays["SHCOARSE"],
        name="Epoch",
        dims=["Epoch"],
        attrs=cdf_utils.epoch_attrs,
    )`

greglucas commented on imap_processing/swe/l1a/swe_l1a.py:
```diff
@@ -13,15 +21,51 @@ def swe_l1a(packet_file: str):
+
+        current_dir = Path(__file__).parent
+        # write data to CDF
+        if apid == SWEAPID.SWE_APP_HK.value:
+            return write_to_cdf(
+                data,
+                "swe",
+                "l1a",
+                version=__version__,
+                mode=f"{data['APP_MODE'].data[0]}",
+                description=get_descriptor(apid),
+                directory=current_dir,
+            )
+        else:
+            return write_to_cdf(
+                data,
+                "swe",
+                "l1a",
+                version=__version__,
+                description=get_descriptor(apid),
+                directory=current_dir,
+            )
```
Put the only difference into an if statement and return after to shorten the implementation?

```suggestion
      return write_to_cdf(
          data,
          "swe",
          "l1a",
          version=__version__,
          mode=f"{data['APP_MODE'].data[0]}" if apid == SWEAPID.SWE_APP_HK else "",
          description=get_descriptor(apid),
          directory=current_dir,
      )
```

greglucas commented on imap_processing/swe/tests/test_swe_l1b.py:
```diff
@@ -0,0 +1,167 @@
+    ]
+    evtmsg_l1b_filepath = swe_l1b(evtmsg_data[0])
+
+    assert os.path.basename(sci_l1b_filepath) == "imap_swe_l1b_20230927_sci_v01.cdf"
```
Are these `Path` objects? if so, you can just get the name. If not, maybe you can make the return be a path object instead of a string path?
```suggestion
    assert sci_l1b_filepath.name == "imap_swe_l1b_20230927_sci_v01.cdf"
```

tech3371 replied:
It looks like it's not `Path` object but a string.

tech3371 committed:
feedback changes

tech3371 committed:
feedback changes.

tech3371 committed:
merged write_to_cdf.py into cdf_utils.py. Made updates to other files to import that and other updates as needed. WIP: need to fix more error from this change. Will get back to it

tech3371 committed:
removed evtmsg test since logic has changed

bourque commented on imap_processing/swe/l1b/swe_l1b_science.py:
```diff
@@ -0,0 +1,485 @@
+    Deadtime correction is important to ensure accurate measurements and data
+    analysis in fields where event detection rates are high and where every
+    detected event is critical for understanding physical processes.
+
+
```
Some extra lines here

bryan-harter commented on imap_processing/swe/l1a/swe_science.py:
```diff
@@ -130,50 +118,81 @@ def swe_science(decom_data):
+    # TODO: refactor this better once we develop class inheritance
+    int_attrs = swe_cdf_attrs.int_attrs
+    int_attrs["CATDESC"] = int_attrs["FIELDNAM"] = int_attrs["LABLAXIS"] = "Energy"
+    int_attrs["VALIDMAX"] = np.int64(180)
```
I don't think it hurts anything to have VALIDMIN and VALIDMAX here, but they aren't actually required if these numbers don't change over time

tech3371 replied:
oh ok. good to know

bourque commented on imap_processing/swe/l1b/swe_l1b_science.py:
```diff
@@ -0,0 +1,485 @@
+
+
+def deadtime_correction(counts: np.array, tsampl: int):
+    """Calculate deadtime correction.
+
+    Deadtime correction is a technique used in various fields, including
+    nuclear physics,radiation detection, and particle counting, to compensate
+    for the effects of the time period during which a detector is not able to
+    record new events or measurements after detecting a previous event.
+    This "deadtime" is essentially the time during which the detector is
+    recovering from the previous detection and is unable to detect new events.
+
+    In particle detectors, there is a finite time required for the detector to
+    reset or recover after detecting a particle. During this deadtime, any
+    subsequent particles that may have arrived go undetected. As a result,
+    the recorded count rate appears to be lower than the actual count rate.
+
+    Deadtime correction involves mathematically adjusting the measured count
+    rates to compensate for this deadtime effect. This correction is crucial
+    when dealing with high-intensity sources or particle fluxes, as the deadtime
+    can significantly affect the accuracy of the measurements.
+
+    Deadtime correction is important to ensure accurate measurements and data
+    analysis in fields where event detection rates are high and where every
+    detected event is critical for understanding physical processes.
```
I like how you included this detailed description right in the code!

bryan-harter commented on imap_processing/swe/l1a/swe_science.py:
```diff
@@ -130,50 +118,81 @@ def swe_science(decom_data):
+
+    # TODO: refactor this better once we develop class inheritance
+    int_attrs = swe_cdf_attrs.int_attrs
+    int_attrs["CATDESC"] = int_attrs["FIELDNAM"] = int_attrs["LABLAXIS"] = "Energy"
```
SPDF is definitely gonna want something a bit more descriptive, roughly 80 characters for CATDESC, 30 for FIELDNAM.  I think LABLAXIS is fine with just "energy".

tech3371 replied:
Totally agree. I plan to add more descriptive description in future PR. This PR already had lots of changes. I will create issue for that upcoming PR though.

bourque commented on imap_processing/swe/l1b/swe_l1b_science.py:
```diff
@@ -0,0 +1,485 @@
+    ----------
+    counts : np.array
+        counts data before deadtime corrections
+    tsampl : int
```
This parameter name is a bit confusing to me. Is it supposed to represent "time sample"?. Perhaps call it `time_sample`? or `acq_duration`? or just `duration`?

tech3371 replied:
I like `acq_duration`

bryan-harter commented on imap_processing/swe/l1a/swe_science.py:
```diff
@@ -130,50 +118,81 @@ def swe_science(decom_data):
+        int_attrs["CATDESC"] = int_attrs["FIELDNAM"] = int_attrs["LABLAXIS"] = key
+        # get int32's max since most of metadata is under 32-bits
+        int_attrs["VALIDMAX"] = np.iinfo(np.int32).max
+        int_attrs["DEPEND_O"] = "Epoch"
```
Should be a DEPEND_0 (zero), not DEPEND_O (the letter)

tech3371 replied:
good catch!

bryan-harter commented on imap_processing/swe/swe_cdf_attrs.py:
```diff
@@ -0,0 +1,227 @@
+# value.
+int_attrs = {
+    "CATDESC": None,
+    "DISPLAY_TYPE": "spectrogram",
```
The only variables that should have a DISPLAY_TYPE are the data variables.  Support_data doesn't need it, and SPDF might comment on it.

tech3371 replied:
Good to know. This will probably go away when I rebase with Maxine's changes.

bourque commented on imap_processing/utils.py:
```diff
@@ -0,0 +1,94 @@
+import collections
```
It would be good to add a module docstring here, especially since it will likely be used by the whole team

tech3371 replied:
Sure thing

bourque reviewed (approved):
Awesome work here @tech3371! I just had a few comments, no blockers. Thanks for writing all of this; it will also be very helpful for CoDICE decom and testing.

tech3371 committed:
feedback changes. TODO: rebase with Maxine's and make changes

tech3371 committed:
Initial changes to change from hard coded dictionary to use dataclasses objects. WIP

tech3371 committed:
second half of refactor to use CDF dataclasses

greglucas commented on imap_processing/swe/l1b/swe_l1b_science.py:
```diff
@@ -0,0 +1,509 @@
+    table_index_value : int
+        ESA table index number
+    """
+    lookup_table_filepath = f"{imap_module_directory}/swe/l1b/swe_esa_lookup_table.csv"
```
Not critical, but you could keep it as a Path object here instead of string.
```suggestion
    lookup_table_filepath = imap_module_directory / "swe/l1b/swe_esa_lookup_table.csv"
```

greglucas commented on imap_processing/swe/l1b/swe_l1b_science.py:
```diff
@@ -0,0 +1,509 @@
+    if table_index_value == 0:
+        return lookup_table.loc[lookup_table["table_index"] == 0]
+
+    if table_index_value == 1:
```
```suggestion
        return lookup_table.loc[lookup_table["table_index"] == 0]
    elif table_index_value == 1:
```

greglucas commented on imap_processing/swe/l1b/swe_l1b_science.py:
```diff
@@ -0,0 +1,509 @@
+    """Calculate deadtime correction.
+
+    Deadtime correction is a technique used in various fields, including
+    nuclear physics,radiation detection, and particle counting, to compensate
```
```suggestion
    nuclear physics, radiation detection, and particle counting, to compensate
```

greglucas commented on imap_processing/swe/l1b/swe_l1b_science.py:
```diff
@@ -0,0 +1,509 @@
+    np.array
+        Count rates array in seconds
+    """
+    return np.divide(data, acq_duration)
```
Out of curiosity is there a reason to use `np.divide(a, b)` instead of `a / b`?

tech3371 replied:
no particular reason. I feel like I heard in the past that using np.divide is suggested.

greglucas replied:
The only reason I'd possibly suggest avoiding it is because of some of the deferral mechanisms for other array libraries. i.e. if you do xarray.Dataset / np.ma.MaskedArray, using `np.divide()` can force those to numpy-arrays rather than letting the dataset and masked array choose whether to return a Dataset with metadata still attached instead. No concern with what you have here, but I would actually suggest the opposite in a more general sense.

tech3371 replied:
That's good to know. ty!

greglucas commented on imap_processing/swe/l1b/swe_l1b_science.py:
```diff
@@ -0,0 +1,509 @@
+
+        # Go through four quarter cycle data packets
+        for index in range(4):
+            uncompressed_counts = l1a_data["SCIENCE_DATA"].data[packet_index + index]
```
I think we decided on decompressed for consistency.
```suggestion
            decompressed_counts = l1a_data["SCIENCE_DATA"].data[packet_index + index]
```

greglucas commented on imap_processing/swe/l1b/swe_l1b_science.py:
```diff
@@ -0,0 +1,509 @@
+        column_index = -1
+
+        # Go through four quarter cycle data packets
+        for index in range(4):
```
If it turns out this is a bottleneck in the future (no need to update now), these nested for-loops could possibly be refactored out into fancy indexing. I'm sure it wouldn't be super easy to do though either.

greglucas commented on imap_processing/swe/l1b/swe_l1b_science.py:
```diff
@@ -0,0 +1,509 @@
+
+
+def swe_l1b_science(l1a_data):
+    """In this function, we cap.
```
Looks like the description wasn't finished.

greglucas commented on imap_processing/swe/l1b/swe_l1b_science.py:
```diff
@@ -0,0 +1,509 @@
+    xr.dataset
+        L1A dataset with filtered metadata.
+    """
+    for key, value in l1a_data.items():
+        l1a_data[key] = value.data[full_cycle_data_indices]
```
I think you may be able to avoid needing to do any copying in the other function by returning a new dataset here?

You may also be able to index into the entire dataset at once instead of needing to do a for-loop over the individual variables.
```python
return ds.loc[{"dim0": full_cycle_data_indices}]
```
https://docs.xarray.dev/en/latest/user-guide/indexing.html#dataset-indexing

tech3371 replied:
I have some questions about that. Can I revisit this after thanksgiving break? I like to merge in this PR as soon as possible so that others can use common utils from this PR. I appreciate all the suggestions. I learned a lot in this PR!

greglucas replied:
Yes, that is no problem to address it later.

greglucas commented on imap_processing/swe/l1b/swe_l1b_science.py:
```diff
@@ -0,0 +1,509 @@
+            catdesc="Energy",
+            fieldname="Energy",
+            label_axis="Energy",
+            units="keV",
```
Is this the actual energy, or is it the integer index of the energy bin? You put `arange()` in the data field, which makes me think it may be an index and therefore unitless? Same for angle below.

tech3371 replied:
I changed it to this. Does that look ok?
```
energy = xr.DataArray(
        np.arange(180),
        name="Energy",
        dims=["Energy"],
        attrs=dataclasses.replace(
            swe_cdf_attrs.int_base,
            catdesc="Energy's index value in the lookup table",
            fieldname="Energy Bins",
            label_axis="Energy Bins",
            units="int",
        ).output(),
    )
```

greglucas replied:
I don't think "int" is right, it should probably be unitless if that is allowed. Just an empty string maybe.

greglucas replied:
Alternatively, could you actually pass in `decommutation_table` instead of `np.arange(180)` so you got the actual energy values here?

tech3371 replied:
I will leave empty string for now because those values in `decommutation_table` could be different based on look table id. I will keep this mind as we go forward.

greglucas commented on imap_processing/swe/l1b/swe_l1b_science.py:
```diff
@@ -0,0 +1,509 @@
+            catdesc="Rates",
+            fieldname="Rates",
+            label_axis="Rates",
+            units="float",
```
```suggestion
            units="counts/s",
```

greglucas commented on imap_processing/swe/tests/l1a_test_data/imap_swe_l1a_20230927_lveng_hk_v01.cdf:
These cdf files shouldn't be committed to the repo I don't think? It seems like they are leaking out here and we might need to put these in a temporary data test fixture to automatically get cleaned up.

Edit: I just read farther down and I think these are actually the l1a files that go into l1b? I'm hesitant to start committing all intermediate files because I feel like these are going to change and the repository and installed package size will really grow fast... At the same time, you want to be able to run the l1b tests without requiring l1a to run first.

Thoughts on this?

tech3371 replied:
I agree. I was having doubts about that. cool. I will remove it. At least, I got practice on how to read cds files.

greglucas commented on imap_processing/swe/tests/test_swe_l1a.py:
```diff
@@ -0,0 +1,55 @@
+    grouped_data = group_by_apid(decom_test_data)
+    sci_cdf_filepath = swe_l1a(grouped_data[SWEAPID.SWE_SCIENCE])
+    hk_cdf_filepath = swe_l1a(grouped_data[SWEAPID.SWE_APP_HK])
+    assert os.path.basename(sci_cdf_filepath) == "imap_swe_l1a_sci_20230927_v01.cdf"
```
If you return a Path instance from the functions you can use name in the comparison
```suggestion
    assert sci_cdf_filepath.name == "imap_swe_l1a_sci_20230927_v01.cdf"
```

greglucas commented on imap_processing/swe/tests/test_swe_l1a.py:
```diff
@@ -0,0 +1,55 @@
+    assert len(total_event_message_data) == 15
+
+
+def test_cdf_creation(decom_test_data):
```
https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html
```suggestion
def test_cdf_creation(decom_test_data, tmp_path):
```
Then maybe you can call `write_cdf()` with this temporary directory?

tech3371 replied:
I will try that.

tech3371 replied:
changed to use tmp_path

greglucas commented on imap_processing/utils.py:
```diff
@@ -0,0 +1,108 @@
+"""
+Current File contains functions that every instrument can use.
+
+1. sort_by_time function
+    Sorts data packets list by time. It takes packets list
+    and time key as input. Then sorts packets by time key input
+2. group_by_apid function
+    Group packets by apid. It takes packets list as input.
+3. convert_raw_to_eu function
+    Convert raw data to engineering unit. It read conversion
+    table from the given input path and reads current
+    packet name's metadata fields. Then it converts raw data
+    to engineering unit.
```
The functions are listed below with docstrings explaining in more detail what they do so I'd remove the descriptions from the top-level here. I think these should automatically get picked up by the documentation and then we won't get out of sync and need to put information in multiple places.
```suggestion
Common functions that every instrument can use.
```

greglucas reviewed (commented):
Don't be afraid of `Path` objects :) There is a lot of `f"{imap_processing}/path/to/file"` that could be `imap_processing / "path/to/file"` so you stay in the pathlib classes.

The dataset copy / indexing is something worth looking into IMO. I think we should try to return a new view into the subset dataset instead of overwriting the current one if possible.

tech3371 committed:
feedback changes
- renamed to decompressed. fixed comments
- convert everything to Path where possible
- update rates functions
- refactor code to use tmp_path to store test data

tech3371 committed:
changes to use new CDF dataclasses

tech3371 committed:
removed cdf test data. It's now generated at runtime and tests uses latest cdf generated at runtime

greglucas reviewed (approved):
Because you added and removed files, please squash-merge when merging this in.

tech3371 committed:
change energy unit to be unitless for now.
